Stratis.Bitcoin/Utilities/AsyncLock.cs

```diff
@@ -1,59 +1,131 @@
+        }
+
+        /// <summary>Internal synchronization primitive used as a mutex to only allow one thread to occupy the critical section.</summary>
         private readonly SemaphoreSlim semaphore;
+
+        /// <summary>
+        /// Helper object that allows implementation of disposable lock for convenient use with <c>using</c> statement.
+        /// <para>This releaser is used when the lock has been acquired and disposing it will release the lock.</para>
+        /// </summary>
+        /// <remarks>We use the disposable interfaced in a task here to avoid allocations on acquiring the lock when it is free.</remarks>
         private readonly Task<IDisposable> releaser;
 
+        /// <summary>
+        /// Initializes an instance of the object.
+        /// </summary>
         public AsyncLock()
         {
             this.semaphore = new SemaphoreSlim(1, 1);
             this.releaser = Task.FromResult((IDisposable)new Releaser(this));
         }
 
         /// <summary>
-        /// Usage: using(await myAsyncLock.LockAsync())
+        /// Acquires the lock.
         /// </summary>
+        /// <param name="cancel">Cancellation token that can be used to abort waiting for the lock.</param>
+        /// <returns>Disposable interface to enable using construct. Disposing it releases the lock.</returns>
+        /// <exception cref="OperationCanceledException">Thrown when the <paramref name="cancel"/> is triggered and the lock is not acquired.</exception>
         public Task<IDisposable> LockAsync(CancellationToken cancel = default(CancellationToken))
         {
             Task wait = this.semaphore.WaitAsync(cancel);
-
+            
+            // If the lock is available, quickly return.
             if (wait.IsCompleted)
             {
+                // We only hold the lock if the task was completed successfully.
+                // If the task was cancelled, we don't hold the lock and we need to throw.
+                if (wait.IsCanceled) throw new OperationCanceledException();
+
                 return this.releaser;
             }
-            else
+
+            // If the lock is not available, we wait until it is available 
+            // or the wait is cancelled.
+            return wait.ContinueWith((task, state) => 
             {
-                return wait.ContinueWith((_, state) =>
-                    (IDisposable)state,
-                    this.releaser.Result, cancel,
-                    TaskContinuationOptions.ExecuteSynchronously,
-                    TaskScheduler.Default);
-            }
+                // We only hold the lock if the task was completed successfully.
+                // If the task was cancelled, we don't hold the lock and we need to throw.
+                if (task.IsCanceled) throw new OperationCanceledException();
+
+                return (IDisposable)state;
+            }, this.releaser.Result, cancel, TaskContinuationOptions.ExecuteSynchronously, TaskScheduler.Default);
```

Comment:
Is there not an overload without `TaskScheduler.Default`

Reply:
if it delete it, it does not work :) i don't know, this works

Reply:
since the original - https://www.hanselman.com/blog/ComparingTwoTechniquesInNETAsynchronousCoordinationPrimitives.aspx also have it, i doubt that

Reply:
Fair enough yeah sometimes the original just tries to do more then needed.